frontend/public/components/utils/resource-log.jsx

```diff
@@ -0,0 +1,249 @@
+    });
+  }
+
+  // use _updateLogState toggle streaming/paused state instead of directly
+  // setting logState
+  _toggleStreaming() {
+    this._updateLogState(this.state.logState === 'streaming' ? 'paused' : 'streaming');
+  }
+
+  // Set loadtime
+  _touchLoadTime() {
+    this._touchLoadTimeState();
+    this._loadTime = Date.now();
+  }
+
+  // separate function so that it can be throttled
+  _touchLoadTimeState() {
+    this.setState({
+      loadTime: this.state.loadTime + 1
+    });
+  }
+
+  // Abort XMLHttpRequest and update logState accordingly
+  _endStreaming() {
+    // If paused, wait until next change in logState to set to eof
+    if (!this.logState === 'paused') {
+      this._updateLogState('eof');
+    }
+
+    this._stream.abort();
+    clearTimeout(this._pendingReload);
+  }
+
+  // Sets a timeout which will refresh the log from the API every five seconds,
+  // or after the previous load finishes, whichever comes first.
+  _resetPendingReload() {
+    const sinceLastLoad = Date.now() - this._loadTime;
+    const wait = Math.max(0, (1000 * 5) - sinceLastLoad);
+    this._pendingReload = setTimeout(() => {
+      if (this.state.logState === 'paused') {
+        // don't reset stream if the user paused the stream
+        this._touchLoadTime();
+        this._resetPendingReload();
+        return;
+      }
+      this._beginStreaming();
+    }, wait);
+  }
+
+  // Callback which handles the XMLHttpRequest "loadstart" event
+  _loadStarted() {
+    this._updateLogState('streaming');
+  }
+
+  // Callback to process data from the XMLHttpRequest "progress" event
+  _processData(data) {
+    if (dataHasHTML(data)) {
+      this._buffer.push('Logs are currently unavailable');
+    } else if (!dataHasFailureMsg(data)) {
+      this._buffer.push(data);
+    }
+    this._touchLoadTime();
+  }
+
+  // Use stream to load log data asynchronously
+  _beginStreaming() {
+    clearTimeout(this._pendingReload);
+    this._pendingReload = null;
+    this._buffer = lineBuffer(1000);
+
+    this._stream = stream(this.state.logURL, this._loadStarted.bind(this), this._processData.bind(this));
+    this._stream.promise
+      .then(() => {
+
+        // Resource is no longer running/generating new log content, so stop streaming
+        if (this.props.eof){
+          this._endStreaming();
+          return;
+        }
+
+        // If another load isn't still running, start a new one
+        if (!this._pendingReload) {
+          this._resetPendingReload();
+        }
+
+      })
+      .catch((why) => { // Load failed/aborted
+        if (why === 'abort') {
+          return;
+        }
+
+        this._loadStarted();
+        this._buffer.push(`Error: ${why}`);
+        this._touchLoadTime();
+        if (!this._pendingReload) {
+          this._resetPendingReload();
+        }
+      });
+  }
+
+  render() {
+    return <div className="co-m-pane__body">
+      <LogControls {...this.props}
```

Comment:
Use LogControls component instead of doing a bunch of logic to render the elements inline here.